### ISSUE TITLE
Could we not use `Show` but only have instances for basic types?

### ISSUE THREAD

**Issue by i-am-the-slime** [JordanMartinez/purescript-interpolate#2]
So instead of having an instance for anything with show have one for:
`Boolean`, `Int`, `Number`, `Char`. I think using `Show` is almost never what you (or at least I) want and using it for this seems as dangerous as implicit `toString` in Java.
I think this is probably used for crafting specific - pretty - strings.

`(SomeType "hello")` is not very pretty in my book. Further, I don't think it's too painful to write `show xyz` if you actually do want that behaviour.

**Comment by JordanMartinez** [comment]
I don't really see the issue here. If you don't want to use the type's `Show` instance, you could just render the value into a String and go from there
```purescript
i "some text message " (someMaybe # maybe "" \r -> r.textValue) " other text"
```

The only Show instance that could be problematic is `Char` because it adds the single quotations around the character.

**Comment by i-am-the-slime** [comment]
I think this is a matter of preference. I like to get as much help from the compiler as possible and don't mind writing `show` if I really want it. I've seen code where an id generation accidentally included around some newtype because of this implicitness in Scala. So in Haskell and Purescript I had to manually wrap my values in this string before hashing them. I know it's a sample size of one but if Scala didn't just happily `.toString()` on anything that comes its way this wouldn't have happened. At the expense of some unwrapping. `"PartnerId(" <> theActualString <> ")"`

I would make the case that if you *do* want to use the `Show` instance, it's not hard to do either. But you get the compiler to make you stop and think whether you might rather want `unwrap` or something completely different.

**Comment by TheMatten** [comment]
I guess this seems like a `Display`/`Debug` distinction in Rust - `Debug` is used for pretty-printing values in way they are defined (as with `Show`), while `Display` is used for user-facing output - so e.g. `newtype Kilometres a = Kilometres a; Kilometres 42` could be `display`-ed as `42km`. Maybe what you want here is `Display`-like class?

**Comment by JordanMartinez** [comment]
There's a few thoughts I have on this.

@TheMatten Yeah, I think `Display` is more of what we have in mind here: something that is user-facing. I think that's what @i-am-the-slime is getting at above, too. 
Harry already wrote a `Debug`-like type class. See [Harry's 'Down with Show' blog post series](https://harry.garrood.me/blog/down-with-show-part-1/) and the repo he created, [`purescript-debugged`](https://github.com/hdgarrood/purescript-debugged).

@i-am-the-slime has a valid point. While it is more convenient to just use `Show` instances, there are good reasons to avoid it (e.g. instance wars, encouraging better practices, etc.). I think having more explicitness here at the cost of some extra verbosity might be better.

On the other hand, I don't see why there can't be two versions of this type class, one that defaults to using `Show` instances and one that does not, giving end-users the right to choose their own tradeoffs.

**Comment by i-am-the-slime** [comment]
@JordanMartinez @TheMatten `Display` sounds indeed like what I would want (I still wouldn't use `Show` for debugging, I wouldn't use it for anything except turning basic types into strings, but that's a different topic).

@JordanMartinez 

> On the other hand, I don't see why there can't be two versions of this type class, one that defaults to using Show instances and one that does not, giving end-users the right to choose their own tradeoffs.

I'm not sure if two will be enough. Maybe this is a case for vendoring the library in. I personally would have an instance for anything newtyped which might be completely out of line with what I said before, but 🤷🏻‍♂️: 

```purescript
else instance interpNewtypeFunction :: (Newtype n b, Interpolate (b -> a)) => Interpolate (n -> a) where
  interpolate str nt = interpolate str (unwrap nt)
```

**Comment by JordanMartinez** [comment]
> I'm not sure if two will be enough. Maybe this is a case for vendoring the library in. I personally would have an instance for anything newtyped which might be completely out of line with what I said before, but 🤷🏻‍♂️:

I think on that note, you've convinced me to just do the primitive types. Otherwise, we get into instance wars.

**Comment by JordanMartinez** [comment]
Please give me your thoughts on #4

**Comment by JordanMartinez** [comment]
Closed by #4
